File: android/app/src/main/kotlin/com/travelexpenses/android/ui/trips/TripViewModel.kt
package com.travelexpenses.android.ui.trips

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.travelexpenses.currency.CurrencyConverter
import com.travelexpenses.event.EventReplayEngine
import com.travelexpenses.event.ExpenseEvent
import com.travelexpenses.model.Expense
import com.travelexpenses.model.Trip
import com.travelexpenses.model.TripId
import com.travelexpenses.repository.EventLogRepository
import com.travelexpenses.repository.ExpenseRepository
import com.travelexpenses.repository.TripRepository
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import kotlinx.datetime.Clock
import kotlinx.datetime.LocalDate
import java.util.UUID

class TripViewModel(
    private val tripRepo: TripRepository,
    private val expenseRepo: ExpenseRepository,
    private val eventLogRepo: EventLogRepository,
    private val replayEngine: EventReplayEngine,
    private val currencyConverter: CurrencyConverter,
    private val deviceId: String,
) : ViewModel() {

    val activeTrips: StateFlow<List<Trip>> = tripRepo.observeActiveTrips()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    private val _archivedTrips = MutableStateFlow<List<Trip>>(emptyList())
    val archivedTrips: StateFlow<List<Trip>> = _archivedTrips

    private val _selectedTrip = MutableStateFlow<Trip?>(null)
    val selectedTrip: StateFlow<Trip?> = _selectedTrip

    private val _tripExpenses = MutableStateFlow<List<Expense>>(emptyList())
    val tripExpenses: StateFlow<List<Expense>> = _tripExpenses

    private val _tripTotal = MutableStateFlow<String?>(null)
    val tripTotal: StateFlow<String?> = _tripTotal

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading

    init {
        loadArchivedTrips()
    }

    private fun loadArchivedTrips() {
        viewModelScope.launch {
            _archivedTrips.value = tripRepo.getArchivedTrips()
        }
    }

    fun loadTrip(tripId: TripId) {
        viewModelScope.launch {
            _isLoading.value = true
            _selectedTrip.value = tripRepo.getTrip(tripId)
            _tripExpenses.value = expenseRepo.getExpensesForTrip(tripId)
            val trip = _selectedTrip.value
            if (trip != null && _tripExpenses.value.isNotEmpty()) {
                val result = currencyConverter.aggregateTripTotal(
                    _tripExpenses.value,
                    trip.baseCurrency,
                )
                _tripTotal.value = "${result.total} ${trip.baseCurrency}"
            } else {
                _tripTotal.value = null
            }
            _isLoading.value = false
        }
    }

    private suspend fun nextSequenceNumber(): Long = eventLogRepo.count() + 1

    fun createTrip(
        name: String,
        destination: String?,
        baseCurrency: String,
        startDate: LocalDate?,
        endDate: LocalDate?,
    ) {
        viewModelScope.launch {
            val now = Clock.System.now()
            val tripId = UUID.randomUUID().toString()
            val trip = Trip(
                id = tripId,
                name = name,
                destination = destination,
                startDate = startDate,
                endDate = endDate,
                baseCurrency = baseCurrency,
                createdAt = now,
            )
            val event = ExpenseEvent.TripCreated(
                eventId = UUID.randomUUID().toString(),
                timestamp = now,
                sequenceNumber = nextSequenceNumber(),
                deviceId = deviceId,
                trip = trip,
            )
            eventLogRepo.append(event)
            loadArchivedTrips()
        }
    }

    fun archiveTrip(tripId: TripId) {
        viewModelScope.launch {
            val now = Clock.System.now()
            val event = ExpenseEvent.TripArchived(
                eventId = UUID.randomUUID().toString(),
                timestamp = now,
                sequenceNumber = nextSequenceNumber(),
                deviceId = deviceId,
                tripId = tripId,
            )
            eventLogRepo.append(event)
            loadArchivedTrips()
        }
    }

    fun deleteExpense(expenseId: String) {
        viewModelScope.launch {
            val now = Clock.System.now()
            val event = ExpenseEvent.ExpenseDeleted(
                eventId = UUID.randomUUID().toString(),
                timestamp = now,
                sequenceNumber = nextSequenceNumber(),
                deviceId = deviceId,
                expenseId = expenseId,
            )
            eventLogRepo.append(event)
            _selectedTrip.value?.id?.let { loadTrip(it) }
        }
    }
}

File: ios/TravelExpenses/UI/Trips/TripViewModel.swift
import Foundation
import Shared

@MainActor
final class TripViewModel: ObservableObject {

    @Published var activeTrips: [Trip] = []
    @Published var archivedTrips: [Trip] = []
    @Published var selectedTrip: Trip?
    @Published var tripExpenses: [Expense] = []
    @Published var tripTotal: String = "0.00"
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?

    private let tripRepo: SqlDelightTripRepository
    private let expenseRepo: SqlDelightExpenseRepository
    private let eventLogRepo: SqlDelightEventLogRepository
    private let currencyConverter: CurrencyConverter
    private let deviceId: String

    // Flow bridges (keep alive while ViewModel is alive)
    private var tripBridge: TripFlowBridge?
    private var expenseBridge: ExpenseFlowBridge?

    init(container: ServiceContainer) {
        self.tripRepo       = container.tripRepo
        self.expenseRepo    = container.expenseRepo
        self.eventLogRepo   = container.eventLogRepo
        self.currencyConverter = container.currencyConverter
        self.deviceId       = container.deviceId

        observeTrips()
    }

    // MARK: - Flow observation

    private func observeTrips() {
        let bridge = TripFlowBridge(repo: tripRepo)
        self.tripBridge = bridge

        bridge.collectActiveTrips(
            onEach: { [weak self] trips in
                Task { @MainActor in
                    self?.activeTrips = trips as? [Trip] ?? []
                }
            },
            onError: { [weak self] error in
                Task { @MainActor in self?.errorMessage = error.localizedDescription }
            }
        )

        bridge.collectAllTrips(
            onEach: { [weak self] trips in
                Task { @MainActor in
                    let all = trips as? [Trip] ?? []
                    self?.archivedTrips = all.filter { $0.isArchived == true }
                }
            },
            onError: { _ in }
        )
    }

    // MARK: - Trip CRUD

    func createTrip(
        name: String,
        destination: String?,
        baseCurrency: String,
        startDate: Kotlinx_datetimeLocalDate?,
        endDate: Kotlinx_datetimeLocalDate?
    ) {
        Task {
            do {
                let seqNum = try await eventLogRepo.count() + 1
                let now = Kotlinx_datetimeClock.companion.System.now()
                let trip = Trip(
                    id: UUID().uuidString,
                    name: name,
                    destination: destination,
                    startDate: startDate,
                    endDate: endDate,
                    baseCurrency: baseCurrency,
                    createdAt: now
                )
                let event = ExpenseEvent.TripCreated(
                    eventId: UUID().uuidString,
                    timestamp: now,
                    sequenceNumber: seqNum,
                    deviceId: deviceId,
                    trip: trip
                )
                try await eventLogRepo.append(event: event)
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    func archiveTrip(_ tripId: String) {
        Task {
            do {
                let seqNum = try await eventLogRepo.count() + 1
                let now = Kotlinx_datetimeClock.companion.System.now()
                let event = ExpenseEvent.TripArchived(
                    eventId: UUID().uuidString,
                    timestamp: now,
                    sequenceNumber: seqNum,
                    deviceId: deviceId,
                    tripId: tripId
                )
                try await eventLogRepo.append(event: event)
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    // MARK: - Trip detail

    func loadTripDetail(tripId: String) {
        Task {
            isLoading = true
            do {
                selectedTrip = try await tripRepo.getTrip(tripId: tripId)

                // Start observing expenses for this trip
                let bridge = ExpenseFlowBridge(repo: expenseRepo)
                self.expenseBridge = bridge
                bridge.collectForTrip(
                    tripId: tripId,
                    onEach: { [weak self] expenses in
                        Task { @MainActor in
                            self?.tripExpenses = expenses as? [Expense] ?? []
                            self?.recalculateTotal()
                        }
                    },
                    onError: { _ in }
                )
            } catch {
                errorMessage = error.localizedDescription
            }
            isLoading = false
        }
    }

    func deleteExpense(_ expenseId: String) {
        Task {
            do {
                let seqNum = try await eventLogRepo.count() + 1
                let now = Kotlinx_datetimeClock.companion.System.now()
                let event = ExpenseEvent.ExpenseDeleted(
                    eventId: UUID().uuidString,
                    timestamp: now,
                    sequenceNumber: seqNum,
                    deviceId: deviceId,
                    expenseId: expenseId
                )
                try await eventLogRepo.append(event: event)
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    // MARK: - Private

    private func recalculateTotal() {
        guard let trip = selectedTrip else { return }
        Task {
            do {
                let result = try await currencyConverter.aggregateTripTotal(
                    expenses: tripExpenses,
                    baseCurrency: trip.baseCurrency
                )
                tripTotal = "\(result.total) \(trip.baseCurrency)"
            } catch {
                tripTotal = "—"
            }
        }
    }
}
